File: examples/window.rs
//! Open a window and animate it, driven entirely from Rust.
//!
//! Run with:
//!
//!     cargo run --example window --features "window,metal"
//!
//! Swap `metal` for `vulkan` on Linux and Windows. The `window` feature is
//! what makes any of this compile; without a GPU backend there is nothing to
//! present a frame with.
//!
//! Escape or closing the window ends the loop.

use std::{cell::Cell, rc::Rc};

use meo_skia_canvas::prelude::*;

fn main() {
    let mut win = Window::new(480.0, 320.0);
    win.set_title("meo-skia-canvas");
    win.set_background("#101014");

    // Follows the pointer, so it is visibly reacting to events rather than
    // just animating on a timer.
    //
    // Shared through an `Rc<Cell<_>>` because the two handlers are separate
    // closures: each `move` would otherwise capture its own copy of a plain
    // local, and the one the draw handler read would never be the one the
    // event handler wrote.
    let pointer = Rc::new(Cell::new(Point::new(240.0, 160.0)));

    let seen = Rc::clone(&pointer);
    win.on_event(move |event| {
        if let UiEvent::Mouse { point, .. } = event {
            seen.set(*point);
        }
    });

    win.on_draw(move |ctx, frame| {
        let pointer = pointer.get();
        ctx.set_fill_style(RgbaLinear::opaque(0.06, 0.06, 0.08));
        ctx.fill_rect(0.0, 0.0, 480.0, 320.0);

        // The frame counter is handed in rather than tracked here, so the
        // animation does not need any state of its own.
        let phase = frame as f32 / 30.0;
        let radius = 40.0 + 12.0 * phase.sin();

        ctx.set_fill_style_css("skyblue").ok();
        ctx.begin_path();
        ctx.arc(
            pointer.x,
            pointer.y,
            radius,
            0.0,
            std::f32::consts::TAU,
            false,
        )
        .ok();
        ctx.fill(FillRule::NonZero);
    });

    win.open();

    // Blocks until the last window closes. On macOS this has to be the main
    // thread, which is why it is called here rather than from a worker.
    App::run();
}

File: examples/node/window.js
//
// A window, animated and driven by events
//
// Run from a checkout:
//
//     just build            # or: npm run build
//     node examples/node/window.js
//
// The Node half of `examples/rust/window.rs`. Same drawing, same two
// handlers, so the files read as a translation of each other.
//
// Needs a display. This is the one example that cannot run headless: it opens
// a real window and blocks until it is closed.
//
// The require below is relative because the repo is not linked to itself.
// In your own project it is:  require("meo-skia-canvas")
//

const { Window } = require("../../lib");

const W = 480,
  H = 320;

const win = new Window(W, H, {
  title: "meo-skia-canvas",
  background: "#101014",
});

// Follows the pointer, so it is visibly reacting to events rather than just
// animating on a timer.
//
// A plain closure variable is enough here: JavaScript closures share one
// binding, so both handlers see the same `pointer`. The Rust mirror needs an
// `Rc<Cell<_>>` for this, because each `move` closure would otherwise capture
// its own copy.
let pointer = { x: W / 2, y: H / 2 };

win.on("mousemove", ({ x, y }) => {
  pointer = { x, y };
});

// Escape closes the window, which ends the loop once it is the last one.
win.on("keydown", ({ key }) => {
  if (key === "Escape") win.close();
});

win.on("draw", ({ frame }) => {
  const ctx = win.ctx;
  ctx.fillStyle = "#0f0f14";
  ctx.fillRect(0, 0, W, H);

  // The frame counter is handed in rather than tracked here, so the animation
  // does not need any state of its own.
  const phase = frame / 30;
  const radius = 40 + 12 * Math.sin(phase);

  ctx.fillStyle = "skyblue";
  ctx.beginPath();
  ctx.arc(pointer.x, pointer.y, radius, 0, Math.PI * 2);
  ctx.fill();
});
